src/goto-symex/symex_complexity.cpp

```diff
@@ -0,0 +1,229 @@
+  }
+
+  return false;
+}
+
+bool symex_complexityt::are_loop_children_too_complicated(
+  call_stackt &current_call_stack)
+{
+  // Check that current loops don't violate complexity boundaries, if they
+  // do then add that loop to the blacklist for all currently-running loops.
+
+  // We want to make sure that if a certain execution of a loop is too complex
+  // no matter the branching, that we cancel out early. This is critical because
+  // otherwise we will continue to execute nested loops that we already know
+  // are far too complicated to deal with, wasting huge amounts of time.
+  std::size_t sum_complexity = 0;
+
+  // If non-empty, is the lowest-level loop we should blacklist in the stack.
+  irep_idt blacklist;
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+  {
+    std::vector<loop_complexity_infot> info_vectors;
+    for(auto &loop_iter : frame_iter->loop_iterations)
+    {
+      if(loop_iter.second.is_active)
+        info_vectors.emplace_back(loop_iter.first, &loop_iter.second);
+    }
+
+    // We need to execute inside-out in regards to nested loops, so order as
+    // such.
+    std::sort(
+      info_vectors.begin(),
+      info_vectors.end(),
+      [](loop_complexity_infot &loop_a, loop_complexity_infot &loop_b) {
+        return loop_a.loop_info->nesting_level >
+               loop_b.loop_info->nesting_level;
+      });
+
+    for(loop_complexity_infot &loop_comp : info_vectors)
+    {
+      // If we've already violated complexity, add to all parent loops
+      // our currently blacklisted ones
+      if(!blacklist.empty())
+      {
+        loop_comp.loop_info->blacklisted_loops.emplace(blacklist);
+      }
+      else
+      {
+        sum_complexity += loop_comp.loop_info->children_too_complex;
+        if(sum_complexity > max_loops_complexity)
+          blacklist = loop_comp.loop_id;
+      }
+    }
+  }
+
+  return !blacklist.empty();
+}
+
+complexity_violationt
+symex_complexityt::check_complexity(goto_symex_statet &state)
+{
+  if(!complexity_limits_active() || state.guard.is_false())
+    return None;
+
+  std::size_t complexity = state.complexity();
+  if(complexity == 0)
+    return None;
+
+  auto &current_call_stack = state.call_stack();
+
+  // Check if this branch is too complicated to continue.
+  optionalt<loop_complexity_infot> active_loop =
+    get_current_active_loop(current_call_stack);
+  if(complexity >= max_complexity)
+  {
+    // If we're too complex, add a counter to the current loop we're in and
+    // check if we've violated children complexity limits.
+    if(active_loop)
+    {
+      // This is actually used, don't remove.
+      active_loop->loop_info->children_too_complex++;
+
+      // If we're considered too complex, cancel branch.
+      if(are_loop_children_too_complicated(current_call_stack))
+      {
+        log.warning()
+          << "[symex-complexity] Loop operations considered too complex"
```

Comment:
Mention in the log message what we're doing, not just what we observed (e.g. "skipping", "dropping path", "assuming not entered / taken"